mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -408,9 +409,11 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
    */
   private def parallelPoolAdjacentViolators(
       input: RDD[(Double, Double, Double)]): Array[(Double, Double, Double)] = {
-    val parallelStepResult = input
-      .sortBy(x => (x._2, x._1))
-      .glom()
+    val keyedInput = input.keyBy(_._2)
+    val parallelStepResult = keyedInput
+      .partitionBy(new RangePartitioner(keyedInput.getNumPartitions, keyedInput))
+      .values
+      .mapPartitions( p => Iterator(p.toSeq.sortBy(x => (x._2, x._1)).toArray))
```

Comment:
Actually, the `Iterator` is necessary. Since we removed the `glom`, we need to return an `RDD[Array[T]]`. So the function passed to `mapPartitions` needs to be `Iterator[T] => Iterator[Array[T]]`. Removed the extraneous `toSeq`, though.